File: algorithms/algorithms-cplusplus/leetcode/Question_0045_Jump_Game_II.cc
#include <vector>

using namespace std;

class Solution {
 public:
  int jump(vector<int>& nums) {
    int res = 0, n = nums.size(), i = 0, cur = 0, pre = 0;
    while (cur < n - 1) {
      ++res;
      pre = cur;
      for (; i <= pre; i++) {
        cur = max(cur, i + nums[i]);
      }
      if (pre == cur) {
        return -1;
      }
    }
    return res;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0045_Jump_Game_II.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res, n, last, cur = 0, len(nums), 0, 0
        for i in range(n - 1):
            cur = max(cur, i + nums[i])
            if i == last:
                last = cur
                res += 1
                if cur >= n - 1:
                    break
        return res
